component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/SchemaImpl.java

```diff
@@ -22,46 +22,62 @@
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Stream;
 
 import javax.json.bind.annotation.JsonbTransient;
 
 import org.talend.sdk.component.api.record.Schema;
 
-import lombok.AllArgsConstructor;
 import lombok.Data;
-import lombok.NoArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
 
-@Data
-@NoArgsConstructor
-@AllArgsConstructor
+@EqualsAndHashCode
 public class SchemaImpl implements Schema {
 
-    private Type type;
+    @Getter
+    private final Type type;
 
-    private Schema elementSchema;
+    @Getter
+    private final Schema elementSchema;
 
-    private List<Entry> entries;
+    @Getter
+    private final List<Entry> entries;
 
+    @JsonbTransient
+    private final List<Entry> metadataEntries;
+
+    @Getter
     private Map<String, String> props = new LinkedHashMap<>(0);
 
-    public SchemaImpl(final Type type, final Schema schema, final List<Entry> entries) {
-        this.type = type;
-        elementSchema = schema;
-        this.entries = entries;
+    SchemaImpl(final SchemaImpl.BuilderImpl builder) {
+        this.type = builder.type;
+        this.elementSchema = builder.elementSchema;
+        this.entries = unmodifiableList(builder.entries);
+        this.metadataEntries = unmodifiableList(builder.metadataEntries);
+        this.props = builder.props;
     }
 
     @Override
     public String getProp(final String property) {
         return props.get(property);
     }
 
+    @Override
+    @JsonbTransient
+    public Stream<Entry> getAllEntries() {
+        return Stream.concat(this.metadataEntries.stream(), this.entries.stream());
+    }
+
     public static class BuilderImpl implements Builder {
 
         private Type type;
 
         private Schema elementSchema;
 
-        private List<Entry> entries = new ArrayList<>();
+        private final List<Entry> entries = new ArrayList<>();
+
+        private final List<Entry> metadataEntries = new ArrayList<>();
```

Comment:
Why use 2 lists?

Reply:
The method "List<Entry> getEntries();" of Schema interface already exist and is widely used by our connectors.
Using one list would lead to rewrite this method with "entries.stream().filter( !metadata).collect(Collectors.toList())" and so reconstruct a new List and degrade performance.

Reply:
see my comment below... Schema is not handled on each record (in most cases...) so performance should not be an issue in that case...

Reply:
?? each record has a schema, [see RecordImpl.BuilderImpl.build()](https://github.com/Talend/component-runtime/blob/master/component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java#L273-L274) method.
So, even if a record has no metadata; puting list of data and meta data together will oblige to build a new list on each call of getEntries() method (this.entries.stream().filter(not meta)...) because most of component must not see metadata (it must be transparent until you asked explicitly for it)). So; performance issue would be always here.

Moreover, having a field name "entries" and a method "getEntries" that doest not reflect the list content is perturbing; or, at least, with this solution, i can rename it to allEntries; but performance issue will be here for ALL records even those without metadata.